drivers/sensor/st/stm32_dts/stm32_dts.c

```diff
@@ -0,0 +1,283 @@
+/*
+ * Copyright (c) 2024 Aurelien Jarno
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT st_stm32_dts
+
+#include <zephyr/device.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(stm32_dts, CONFIG_SENSOR_LOG_LEVEL);
+
+/* Constants */
+#define ONE_MHZ			1000000 /* Hz */
+#define TS1_T0_VAL0		30	/* °C */
+#define TS1_T0_VAL1		130	/* °C */
+#define SAMPLING_TIME		15	/* best precision */
+
+struct stm32_dts_data {
+	struct k_sem sem_isr;
+	struct k_mutex mutex;
+
+	/* Peripheral clock frequency */
+	uint32_t pclk_freq;
+	/* Engineering value of the frequency measured at T0 in Hz */
+	uint32_t t0_freq;
+	/* Engineering value of the T0 temperature in °C */
+	uint16_t t0;
+	/* Engineering value of the ramp coefficient in Hz / °C */
+	uint16_t ramp_coeff;
+
+	/* Raw sensor value */
+	uint16_t raw;
+};
+
+struct stm32_dts_config {
+	/* DTS instance. */
+	DTS_TypeDef *base;
+	/* Clock configuration. */
+	struct stm32_pclken pclken;
+};
+
+static void stm32_dts_isr(const struct device *dev)
+{
+	struct stm32_dts_data *data = dev->data;
+	const struct stm32_dts_config *cfg = dev->config;
+
+	/* Clear interrupt */
+	cfg->base->ICIFR |= DTS_ICIFR_TS1_CITEF;
+
+	/* Give semaphore */
+	k_sem_give(&data->sem_isr);
+}
+
+static int stm32_dts_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	const struct stm32_dts_config *cfg = dev->config;
+	struct stm32_dts_data *data = dev->data;
+
+	if (chan != SENSOR_CHAN_ALL && chan != SENSOR_CHAN_DIE_TEMP) {
+		return -ENOTSUP;
+	}
+
+	k_mutex_lock(&data->mutex, K_FOREVER);
+
+	/* Wait for the sensor to be ready (~40µS delay after enabling it) */
+	while ((cfg->base->SR & DTS_SR_TS1_RDY) == 0) {
+		k_yield();
+	}
+
+	/* Trigger a measurement */
+	cfg->base->CFGR1 |= DTS_CFGR1_TS1_START;
```

Comment:
Is there a HAL that could abstract all these direct register operations away from this sensor driver?

Reply:
There is a HAL, but that will make the driver way more complicated, and things like power management would be more difficult to implement.

OTOH, I wish there is a LL like for other drivers, but it is unfortunately not the case. I guess the driver can be converted later if/when it get added to stm32cube.

Reply:
Another drawback of the HAL, is that it only returns an integer temperature.

Reply:
I have just opened https://github.com/STMicroelectronics/stm32h7xx_hal_driver/issues/56